crates/network/src/lib.rs

```diff
@@ -0,0 +1,386 @@
+use std::{fmt::Debug, net::SocketAddr, sync::Arc, time::Duration};
+
+use bevy_app::{App, Plugin, PostStartup, PreUpdate, Update};
+use bevy_ecs::{
+    event::EventWriter,
+    prelude::{Commands, Component, Entity, Event, EventReader, Query, Res, World},
+    system::{Resource, SystemState},
+};
+use bytes::{Bytes, BytesMut};
+use flume::{Receiver, Sender, TryRecvError, TrySendError};
+use rjacraft_protocol::{
+    packets::client::{ClientHandshakePacket, ClientStatusPacket, HandshakeState},
+    Decoder, Encoder,
+};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream, ToSocketAddrs},
+    runtime::Runtime,
+    task::JoinHandle,
+    time::Instant,
+};
+
+use crate::{decode::PacketDecoder, encode::PacketEncoder};
+
+mod decode;
+mod encode;
+
+pub struct NetworkPlugin<A>(pub A);
+
+impl<A> Plugin for NetworkPlugin<A>
+where
+    A: ToSocketAddrs + Clone + Send + Sync + 'static,
+{
+    fn build(&self, app: &mut App) {
+        let runtime = Runtime::new().unwrap();
+
+        let (new_connections_send, new_connections_recv) = flume::bounded::<RemoteConnection>(64);
+        let shared = SharedNetworkState(Arc::new(SharedNetworkStateInner {
+            new_connections_send,
+            new_connections_recv,
+        }));
+
+        app.insert_resource(shared.clone());
+
+        let addr = self.0.clone();
+        let accept_loop_system = move |shared: Res<SharedNetworkState>| {
+            let addr = addr.clone();
+            let _guard = runtime.handle().enter();
+            tokio::spawn(accept_loop(addr, shared.clone()));
+        };
+
+        let spawn_new_connections = move |world: &mut World| {
+            while let Ok(connection) = shared.0.new_connections_recv.try_recv() {
+                world.spawn(connection);
+            }
+        };
+
+        app.add_event::<PacketReceivedEvent>()
+            .add_event::<DisconnectEvent>()
+            .add_event::<HandshakeEvent>()
+            .add_event::<ClientStatusRequestEvent>()
+            .add_systems(PostStartup, accept_loop_system)
+            .add_systems(PreUpdate, spawn_new_connections)
+            .add_systems(Update, (run_packet_event_loop, connection_event_handler));
+    }
+}
+
+async fn accept_loop(addr: impl ToSocketAddrs, shared: SharedNetworkState) {
+    let listener = match TcpListener::bind(addr).await {
+        Ok(listener) => {
+            tracing::info!("Listening on: {}", listener.local_addr().unwrap());
```

Comment:
As for `tracing` macros, it is almost always right to import them for convenience.

Reply:
Aren't the macro names a bit too ambiguous to be used without context?

Reply:
Nope, especially when these are well known names